### ISSUE TITLE
By default just print/output to screen and not into a file

### ISSUE THREAD

**Issue by yarikoptic** [NeurodataWithoutBorders/nwbinspector#76]
## Description

During recent NWB-DANDI dev hackathon, during demonstration of the nwbinspector I forgot to ask -- why does it produce/save log into a file instead of just printing it to the screen (which then could be redirected to a file if desired, or there could be an option `--output-file FILENAME` to make it explicit)?

**Comment by bendichter** [comment]
Yes

**Comment by bendichter** [comment]
It does print them to the screen and also to a log file. The log file is also written to by default

**Comment by CodyCBakerPhD** [comment]
> which then could be redirected to a file if desired, or there could be an option --output-file FILENAME to make it explicit)

Via CLI, option `--log-file-path path_to_file` allows an explicit save path for this. Default is `nwbinspector_log_file.txt` in the working directory.

> During recent NWB-DANDI dev hackathon, during demonstration of the nwbinspector I forgot to ask -- why does it produce/save log into a file instead of just printing it to the screen

In general, the design behind this is that running the full check suite on an NWBFile with a large number of objects can take a while, so the default strategy is to always save the report first; then, the output printed to console is actually just the contents of the file read-out (and optionally colored).

**Comment by yarikoptic** [comment]
Imagine every tool (ls, git,...) dumped its log file right in the directory of invocation by default (says Yarik, the author of duecredit)? Sure thing to you, authors of the tool, it sounds like a good idea, but IMHO it is a bad and non conventional default. 
- it might interfere with standardized layouts (eg bids)
- would give impression that may be currently saved listing is up to date when it isn't
- wouldn't work if no write permission
Just my 1c

**Comment by CodyCBakerPhD** [comment]
> wouldn't work if no write permission

What are some common situations for this?

> it might interfere with standardized layouts (eg bids)

If someone stored their NWBFile in BIDS structure and then ran the report from the same location as the file, and then didn't clean it up before sharing it?

> would give impression that may be currently saved listing is up to date when it isn't

First version of NWBFile -> run inspector -> make changes to NWBFile -> report is still there but they don't rerun NWBInspector. True, we've been thinking of some ways of to deal with this (even if we didn't save report by default, would be an issue whenever user decides to write it). I'd almost consider it a separate issue, though.

> Imagine every tool (ls, git,...) dumped its log file right in the directory of invocation by default

Let's pick a good tool for comparison; such as `flake8` or `pytest`. Both are capable of scanning/running on relatively large amounts of data fields (lines of code files or execution of functions). And yeah, I guess neither saves a file by default, but can via flags.

_However_, a key difference between our tool and `flake8` is for code styling, IDEs support line-by-line information about violations so the information is constantly disseminated in a report-like fashion. But there's no IDE for writing NWBFiles, so we think it would really help the user to be able to constantly refer back to the report output to know what else they need to fix.

This makes our tool more like `pytest`, where you just run it every time you make a change to your files and if it breaks a test/violation returns the information about that breakage. With `pytest`, you call `pytest --resultlog=path`, which as you point out is not an enabled feature by default (though personally, I do sometimes wish that it was). _However,_ I wouldn't say `pytest` is super user-friendly to people of all programming levels (including none), which is the audience we target more and more as we make the conversion to NWB much more convenient.



@yarikoptic @bendichter What would you think of the following as a compromise?

Option (a): The good news is we now utilize `click` for all CLI stuff; we could prompt the user and ask their permission to save the report after it runs and prints to console, which also draws their attention to the fact. Something like

```
nwbinspector my_directory

-> prints report to console

Would you like to save this report for future reference? (Y/N)
Y
Where would you like to save the report? (user types relative or absolute path)
/home/jovyan/my_reports

Your NWBInspector report has been saved to /home/jovyan/my_reports!
```

**Comment by CodyCBakerPhD** [comment]
Some other options that occurred to me:

Option (b): we define an explicit `cache` location associated with the inspector. That location could be within the package itself by default: for people doing the convenient PyPI install this would then be all self-contained, for devs we can just add the cache name to the `.gitignore`.

This allows us to store a history of report files, which could be useful on its own for examining the frequency of mistakes particular to an individual installation. For this setup though we would want the file naming scheme to correspond more closely to a particular version of the NWBFiles it was ran on (probably hash the UUIDs or something?).

Option (c): Again with `click`, we can have an initial config setup that triggers the first time someone tries to use the package. This initial setup can pre-define things like if they even want to save reports at all, a cache location for saving all reports if so, format in which to save the files, etc.

Let me know what you think. Also happy to schedule one on one meeting to get more feedback.

**Comment by yarikoptic** [comment]
> > wouldn't work if no write permission
> 
> What are some common situations for this?

I decided to check a dataset of someone else.  That is what annoyed me in `git-annex` recently to be unable to do `annex info` which should be read-only operation (per se) on repos of others.  I think that is fixed now ;)

option (a) -- I was also advocate for "more interaction" from tools, but always brought down by good examples e.g. of `git` which doesn't ask questions unless explicitly requested to run in an interactive mode (e.g. `git rebase -i`).   The point is -- you don't want inconsistent behavior between interactive and non-interactive (cron job etc) modes.

> Option (b): we define an explicit `cache` location associated with the inspector. That location could be within the package itself by default: for people doing the convenient PyPI install this would then be all self-contained, for devs we can just add the cache name to the `.gitignore`.
> 
> This allows us to store a history of report files, which could be useful on its own for examining the frequency of mistakes particular to an individual installation.

All of that (history etc) would still be possible if someone desires to keep it (under git or not), so I do not see why it should be a default mode of operation.

Having some canonical location within a dataset is appealing, but since there is no "The dataset" format nwbinspector adheres to, there cannot be 1. autodeduction where is the top directory of the dataset if I run inspector within some subdirectory; 2. name which is good for any dataset format.  Thus I would, again, leave it to the user or library using nwbinspector to provide such location/management of the output from the inspector.  E.g., may be for `dandi-cli` `validate` we do add also option to save validation output in some canonical location (e.g. `.dandi/logs/validate.log` or alike) and it would include output from nwbinspector. 

This example also brings a notion of "tool (I just run inspector from command line) vs library (it is used as a part of another tool -- like dandi-cli)" modes of operation.  We only recently provided support in datalad to tell two apart!  Which then relates to

> Option (c): Again with click, we can have an initial config setup ...

I think it is Ok to make it up to a user of inspector used as a "tool" (not as a library!) to allow for configuration on where and if to store the logs.  You will likely need some config file support anyways (allowing to have it system/user/dataset scopes, e.g. like `git` has ;) ), so allowing to optionally specify in the config where and if logs to be stored sounds like a reasonable idea to me.
